hedera-node/hedera-token-service-impl/src/test/java/com/hedera/node/app/service/token/impl/test/handlers/staking/StakingRewardsHandlerImplTest.java

```diff
@@ -0,0 +1,990 @@
+
+        subject.applyStakingRewards(handleContext);
+
+        final var node1Info = writableStakingInfoState.get(1L);
+
+        assertThat(newBalance).isEqualTo(node1Info.unclaimedStakeRewardStart());
+    }
+
+    @Test
+    void anAccountThatStartedStakingAtCurrentPeriodDoesntUnclaimStakeWhenChangingElection() {
+        final var newBalance = 555L * HBARS_TO_TINYBARS;
+        final var copy = customizeAccount(account, newBalance, -1L, false, false);
+        addToState(Map.of(payerId, copy));
+
+        // Change node, so to trigger rewards
+        writableAccountStore.put(account.copyBuilder().stakedNodeId(0L).build());
+
+        given(handleContext.consensusNow()).willReturn(stakePeriodStartInstant);
+        given(handleContext.writableStore(WritableAccountStore.class)).willReturn(writableAccountStore);
+        given(handleContext.writableStore(WritableStakingInfoStore.class)).willReturn(writableStakingInfoStore);
+
+        subject.applyStakingRewards(handleContext);
+
+        final var node1Info = writableStakingInfoState.get(1L);
+
+        assertThat(0).isEqualTo(node1Info.unclaimedStakeRewardStart());
+    }
+
+    @Test
+    void anAccountThatDeclineRewardsDoesntUnclaimStakeWhenChangingElection() {
+        final var newBalance = 555L * HBARS_TO_TINYBARS;
+        final var copy = customizeAccount(account, newBalance, -1L, true, false);
+        addToState(Map.of(payerId, copy));
+
+        // Change node, so to trigger rewards
+        writableAccountStore.put(
+                writableAccountStore.get(payerId).copyBuilder().stakedNodeId(0L).build());
+
+        given(handleContext.consensusNow()).willReturn(originalInstant);
+        given(handleContext.writableStore(WritableAccountStore.class)).willReturn(writableAccountStore);
+        given(handleContext.writableStore(WritableStakingInfoStore.class)).willReturn(writableStakingInfoStore);
+
+        subject.applyStakingRewards(handleContext);
+
+        final var node1Info = writableStakingInfoState.get(1L);
+
+        assertThat(0).isEqualTo(node1Info.unclaimedStakeRewardStart());
+    }
+
+    //    @Test
+    //    void anAutoCreatedAccountShouldNotHaveStakeStartUpdated() {
+    //        final var newId = AccountID.newBuilder().accountNum(10000000000L).build();
+    //        writableAccountStore.put(givenValidAccountBuilder().accountId(newId).build());
+    //
+    //        given(handleContext.consensusNow()).willReturn(stakePeriodStartInstant);
+    //        given(handleContext.writableStore(WritableAccountStore.class)).willReturn(writableAccountStore);
+    //        given(handleContext.writableStore(WritableStakingInfoStore.class)).willReturn(writableStakingInfoStore);
+    //
+    //        subject.applyStakingRewards(handleContext);
+    //
+    //        assertThat(0).isEqualTo(writableAccountStore.get(newId).stakeAtStartOfLastRewardedPeriod());
+    //    }
+
+    @Test
+    void earningZeroRewardsWithStartBeforeLastNonRewardableStillUpdatesSASOLARP() {
```

Comment:
What is `SASOLARP`?

Reply:
haha `stakeAtStartOfLastRewardedPeriod` 😛

Reply:
Got it, thanks!